### ISSUE TITLE
[Clang] libstdc++’s `std::forward_like` doesn’t compile

### ISSUE THREAD

**Issue by Sirraide** [llvm/llvm-project#101614]
Attempting to use libstdc++’s implementation of `std::forward_like` (https://godbolt.org/z/cGK1bsjzM)
```c++
#include <utility>

void f() {
    (void) std::forward_like<int&&>(1);
}
```
results in this error
```
<source>:4:17: error: function 'forward_like<int &&, int>' with deduced return type cannot be used before it is defined
    4 |     (void) std::forward_like<int&&>(1);
      |                 ^
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/15.0.0/../../../../include/c++/15.0.0/bits/move.h:93:3: note: 'forward_like<int &&, int>' declared here
   93 |   forward_like(_Up&& __x) noexcept
      |   ^
```
whereas the same code compiles just fine if `libc++` is used. GCC of course compiles this just fine with `libstdc++`. 

I haven’t looked into this too much, but it might have something to do with the fact that we treat `std::forward_like` as a builtin, and libstdc++’s declaration uses `decltype(auto)` as the return type of `std::forward_like`, which might be confusing us somehow.

**Comment by frederick-vs-ja** [comment]
The reason was discussed in #64029. The implementation strategy of MSVC STL's `forward_like` changed later due to this.

**Comment by Sirraide** [comment]
I see. I wonder if the libstdc++ devs would be willing to switch to using an explicit return type, or if we need to change our handling of `std::forward_like` to somehow account for this.

**Comment by philnik777** [comment]
CC @jwakely

**Comment by jwakely** [comment]
I.e. make libstdc++ slower to compile with GCC, so clang's non-transparent optimisation works?

**Comment by jwakely** [comment]
I suppose I can make it depend on `__clang__`

**Comment by jwakely** [comment]
It seems to me that Clang should only do this optimization if it works. When it doesn't work, don't try to replace `forward_like_ and just use the perfectly valid function that's actually in the source code.

**Comment by Sirraide** [comment]
Yeah, I personally do feel like we should have *some* solution to that, but I’m also not really familiar enough w/ that optimisation to comment on how easy or complicated that would be...
CC @AaronBallman, @cor3ntin, @zygoloid

**Comment by MitalAshok** [comment]
@jwakely I'm don't know if `decltype(auto)` is valid there. For example: https://godbolt.org/z/jv4Pbdhr1

```c++
#include <utility>
#include <type_traits>

static_assert(std::is_same_v<
    decltype(std::forward_like<int&&, int()>),
    auto(int(&&)()) noexcept -> int(&&)()
>);
```

... Doesn't compile with libstdc++'s `forward_like`. (This doesn't violate the addressable function rules, no address/reference is formed to `forward_like`)

This is a very edge-case, it's probably better for Clang to accept `std::forward_like` declared with placeholder types. I wonder if any of the other standard library builtins crash with placeholder types?

**Comment by philnik777** [comment]
@MitalAshok Given that the addressable function rule is there to prevent such shenanigans, I don't think that's supposed to be valid, regardless of whether the wording technically allows it. FWIW I do think you're forming a reference, since the type of the expression `std::forward_list<int&&, int()>` is in fact a reference to a function. The diagnostic message seems to support this.

**Comment by jwakely** [comment]
Right, if the rules allow that, the rules need to be fixed. It's certainly against the spirit of the rules.

**Comment by jwakely** [comment]
I'll change `forward_like` to use our `__like_t` alias, instead of the alias using `forward_like`. But the existing code is already in GCC 14.1 and 14.2 (released yesterday) so clang probably needs to cope with it anyway.

**Comment by AaronBallman** [comment]
> But the existing code is already in GCC 14.1 and 14.2 (released yesterday) so clang probably needs to cope with it anyway.

Agreed, we're going to have to do something to support this.

**Comment by zygoloid** [comment]
If the problem only exists for users of two point releases of libstdc++, I think the best thing to do is to not recognise functions with deduced return types as builtins. We'll lose the optimization entirely, but just for the short term. I think the other options are:
- Instantiate if we need to deduce a return type. This would also lose a lot of the value of the optimization and probably be more complex to implement.
- Compute the return type ourselves. This would definitely be more complex to implement, and more error-prone.

**Comment by jwakely** [comment]
Determining the return type is the expensive part of `forward_like`, so I'm not sure how much benefit there is from not instantiating the actual function body if you still need to instantiate the `__like_t` / `_ForwardLike` / `_Forward_like_t` metafunction that actually computes the type. MSVC seems to use `conditional_t` for that, which is notoriously expensive (maybe they optimize that magically in the front end to make it cheaper?) but even if you use a cheaper alternative to `conditional_t` you need `remove_reference_t` twice. Maybe I'm wrong and avoiding instantiating the function template is still significant. Maybe that assumes that `remove_reference_t` and something-like-conditional_t are both being optimized nowadays? (which isn't entirely true for GCC ... because we don't allow built-ins like `__remove_reference` to be used at the top-level where they would need to be mangled, you still need to instantiate at least one class template to use it in a return type).

**Comment by zygoloid** [comment]
There's a *lot* of benefit from the optimization even though we still compute the return type the hard way. We avoid instantiating a function body, and we avoid emitting a function body, and we avoid inlining a function body. We generate less code at -O0 and less work for the optimizer when optimizing, and we generate less debug information. This all adds up to being a pretty significant net positive (as I recall, something like 2% code size reduction and a similar reduction in compile time). I don't know how much the return type work costs, but it's mostly paid once per type, whereas the other costs that are removed by the builtin are paid once per call. Also, `forward_like` is a very minor part of the optimization; by far the biggest win comes from `move` (where we already have a builtin to compute the return type), but we apply the same optimization to all the standard library functions that take a pointer or reference and just return it unchanged.

**Comment by vient** [comment]
frederick-vs-ja pointed me to this issue from #131708. So, is it intended for Clang 20 to be not usable with libstdc++ 14?

**Comment by jwakely** [comment]
> frederick-vs-ja pointed me to this issue from [#131708](https://github.com/llvm/llvm-project/issues/131708). So, is it intended for Clang 20 to be not usable with libstdc++ 14 in C++23 mode?

Only 14.1 and 14.2, because `forward_like` in libstdc++ was changed more than 6 months ago.

**Comment by vient** [comment]
For now 14.1 and 14.2 are the only releases in 14.x series, not sure when to expect 14.3 to be released.

**Comment by jwakely** [comment]
Around the same time of year as 13.3 was released, and 12.3, and 11.3, and ...

https://gcc.gnu.org/develop.html#timeline

**Comment by vient** [comment]
Thanks, I tried to look at gcc dev mailing list with no result. From the trend of shifting 2-3 weeks forward every year it seems 14.3 is to be expected in the beginning of June.

**Comment by brevzin** [comment]
In case somebody else comes across this issue in the near future before gcc 14.3's release, clang has an `-fno-builtin-std-forward_like` (which I saw [here](https://github.com/llvm/llvm-project/issues/64029#issuecomment-1646679323)). That appears to work for now.
